### ISSUE TITLE
Bug: `no-new-object` has incorrect advice when `Object` constructor is called with an argument

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#17517]
### Environment

Node version: v20.4.0
npm version: v9.7.2
Local ESLint version: v8.48.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19045

### What parser are you using?

Default (Espree)

### What did you do?

```js
/* eslint no-new-object: "error" */
 
var obj = new Object(2);
assert.equal(typeof obj, "object");
assert.equal(obj * obj, 4);
```

### What did you expect to happen?

No error, or a different error message.

### What actually happened?

> 3:11 The object literal notation {} is preferable. ([no-new-object](https://eslint.org/docs/rules/no-new-object))
 
The advice to use a literal like `{}` is not applicable when the `Object` constructor is called with an argument, as in the above example. According to [the relevant specification](https://tc39.es/ecma262/#sec-object-value), calling `new Object(value)` will convert the provided argument into an object, as specified in the [ToObject conversions table](https://tc39.es/ecma262/#table-toobject-conversions). To my knowledge, this result cannot be achieved using object literal notation.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLW5ldy1vYmplY3Q6IFwiZXJyb3JcIiAqL1xuIFxudmFyIG9iaiA9IG5ldyBPYmplY3QoMik7XG5hc3NlcnQuZXF1YWwodHlwZW9mIG9iaiwgXCJvYmplY3RcIik7XG5hc3NlcnQuZXF1YWwob2JqICogb2JqLCA0KTtcbiIsIm9wdGlvbnMiOnsiZW52Ijp7fSwicnVsZXMiOnt9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSJ9fX0=

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

I am not sure about the fix in this case. One option is to change the message of the problem, and simply recommend to avoid using the object constructor, but at that point I think that the rule would be no longer considered stylistic.

Alternatively, `Object(value)` is semantically equivalent to `new Object(value)`, and not forbidden by any rule, so that could be also an option.

Another solution is to change the current behavior and stop reporting a problem when `new Object()` is called with an argument. This was suggested in a previous issue, but the discussion was not conclusive: #11810.
 
Ideas?

**Comment by mdjermanovic** [comment]
Good questions. This rule was introduced along with `no-new-array`, and per the original issue https://github.com/eslint/eslint/issues/136 and PR https://github.com/eslint/eslint/pull/137, the motivation for these rules was indeed to enforce the use of object/array literals over Object/Array constructors.

`no-new-array` was later replaced by `no-array-constructor` to include calls without `new` (https://github.com/eslint/eslint/issues/501), and the new rule was then updated to allow 1-argument constructions (https://github.com/eslint/eslint/issues/500).

So, following this path, `no-new-object` would be replaced by a new rule `no-object-constructor` (or perhaps `prefer-object-literals`) that would be reporting `new Object()` and `Object()`, but not `new Object(value)` and `Object(value)`.

**Comment by nzakas** [comment]
> So, following this path, no-new-object would be replaced by a new rule no-object-constructor (or perhaps prefer-object-literals) that would be reporting new Object() and Object(), but not new Object(value) and Object(value).

I'm not sure it's worth that much effort for such an edge case, but no objections if that's what the team would like to do.

**Comment by snitin315** [comment]
>So, following this path, no-new-object would be replaced by a new rule no-object-constructor (or perhaps prefer-object-literals) that would be reporting new Object() and Object(), but not new Object(value) and Object(value).

I feel we should do this to keep the behavior consistent with `no-array-constructor`. Should we mark this issue for v9 or introduce a new rule ` no-object-constructor ` and remove `no-new-object` later? I feel we can implement this during the v9 release directly.

**Comment by fasttime** [comment]
> > So, following this path, no-new-object would be replaced by a new rule no-object-constructor (or perhaps prefer-object-literals) that would be reporting new Object() and Object(), but not new Object(value) and Object(value).
> 
> I feel we should do this to keep the behavior consistent with `no-array-constructor`. Should we mark this issue for v9 or introduce a new rule `no-object-constructor` and remove `no-new-object` later? I feel we can implement this during the v9 release directly.

I was also wondering if it would be possible today, as it was back in 2014, to simply fix and rename the existing rule in a single breaking change. The alternative to add a new rule, then deprecate and remove the old rule seems less disruptive, but it would likely require some more effort.

**Comment by nzakas** [comment]
> I was also wondering if it would be possible today, as it was back in 2014, to simply fix and rename the existing rule in a single breaking change.

We can't actually remove a rule without breaking the ecosystem. An alternative (which I think we did before), is to rename the rule and create an alias to the current rule name. So we could have both `no-object-constructor` and `no-new-object`, but they both point to the same rule instance.

In order to do that, we'd have to be okay with both rule names existing and having the same behavior for both. If we want different behavior between the two rules, then we'd need to stick with adding a new rule and deprecating the existing one.

Note: None of this is a breaking change, so there's no need to tie it to v9.

**Comment by mdjermanovic** [comment]
If we want a different behavior that requires a new name for the rule, the most straightforward approach and I believe the least surprising change for everyone would be to add a new rule and deprecate the existing one.

**Comment by nzakas** [comment]
Agreed.

**Comment by snitin315** [comment]
Agreed, marking this as accepted.

**Comment by fasttime** [comment]
Thanks for the feedback everybody, I'll work on this as soon as I have a chance. From the above discussion I gather that (correct me please if I'm wrong):
* The new rule shall be named `no-object-constructor`.
* The `no-new-object` rule shall be deprecated without any other changes (except for the documentation).

**Comment by mdjermanovic** [comment]
> * The new rule shall be named `no-object-constructor`.
> * The `no-new-object` rule shall be deprecated without any other changes (except for the documentation).

Sounds good to me.
